### ISSUE TITLE
subject id mismatch in create_output_df_dict() and other bugs I've encontered.

### ISSUE THREAD

**Issue by nimzodisaster** [FCP-INDI/C-PAC#649]
I have tracked down the issue that has stopped my group analyses from continuing to the create_output_df_dict() method in cpac_group_runner.py

here is the method:

> def create_output_df_dict(output_dict_list, inclusion_list=None):
> 
>     import pandas as pd
> 
>     output_df_dict = {}
> 
>     # unique_resource_id is tuple (resource_id,strat_info)
>     for unique_resource_id in output_dict_list.keys():
> 
>         new_df = pd.DataFrame(output_dict_list[unique_resource_id])
>         # drop whatever is not in the inclusion lists
>         if inclusion_list:
>             new_df = new_df[new_df.Participant.isin(inclusion_list)]
>         # unique_resource_id is tuple (resource_id,strat_info)
>         if unique_resource_id not in output_df_dict.keys():
>             output_df_dict[unique_resource_id] = new_df
>     return output_df_dict

These are some example elements from the inclusionlist:
pprint(inclusion_list)

['107609-100_1',
 '106032-100_2',
 '111405-100',
 '111594-100',
 '108670-100',
 '108750-200',
 '109595-100_2',
 '110873-100',
 '107841-100_1',
 '108792-100',
 '107723-100_1',
 '107937-100_2',
 '107925-100_1',
and so on.
.]

Here are a few elements of the new_df dataframe before dropping those not in inclusion list

pprint(new_df)
                                              Filepath MeanDVARS  \
0    /home/gose66/RS/out/pipeline_proc/110021-100_2...   16.7190   
1    /home/gose66/RS/out/pipeline_proc/107560-100_1...   11.9249   
2    /home/gose66/RS/out/pipeline_proc/108904-100_2...   16.9172   
3    /home/gose66/RS/out/pipeline_proc/111926-100_s...   43.8015   
4    /home/gose66/RS/out/pipeline_proc/111189-100_1...   22.2209   
5    /home/gose66/RS/out/pipeline_proc/102969-100_2...   55.4890   
6    /home/gose66/RS/out/pipeline_proc/110052-100_1...   21.1728   
7    /home/gose66/RS/out/pipeline_proc/108378-100_s...   18.1819   
8    /home/gose66/RS/out/pipeline_proc/109810-100_1...   24.1337   
9    /home/gose66/RS/out/pipeline_proc/108369-100_s...   17.0521   
10   /home/gose66/RS/out/pipeline_proc/110830-100_1...   15.6109   
11   ..                                                 ...       ...   
 

    MeanFD_Jenkinson MeanFD_Power         Participant Series  
0             0.0757       0.1229  110021-100_2_sess1   func  
1             0.0391       0.0663  107560-100_1_sess1   func  
2             0.0643       0.1123  108904-100_2_sess1   func  
3             0.2645       0.3856    111926-100_sess1   func  
4             0.0668       0.1111  111189-100_1_sess1  funcf  
5             0.9740       2.0057  102969-100_2_sess1   func  
6             0.0414       0.0702  110052-100_1_sess1   func  
7             0.0687       0.1603    108378-100_sess1   func  
8             0.0711       0.1286  109810-100_1_sess1   func  
9             0.0332       0.0677    108369-100_sess1   func  
10            0.0567       0.1036  110830-100_1_sess1   func  
...               ...          ...                 ...    ...  
  
[300 rows x 6 columns]

You will note that in new_df the Participant string has '_sess1' appended after the SUBID, but not the Participant IDs in the inclusion list. Thus all the subjects get dropped and the pipeline ends.

 My functional data is organized thusly,  /SITE/110830-100_1/sess1/func/funcf.nii.gz 
 where 110830-100_1 is an example Subject.  sess1 is a dummy session; there was only 1 session (sess1) for each subject. This organization was recommended to me to get around a bug in the subject list builder for the requirement of a session directory.
 
 However, in my output directory, all subjects have the '_sess1' appended to their IDs, like this:
 
 /home/gose66/RS/out/pipeline_proc/110830-100_1_sess1
 
 So now it seems I have the awkward situation of having subject IDs in my output not matching the subject IDs in my data inputs, and I'm not sure how to proceed, or the code is not stripping the session label from the subject IDs.

**Comment by nimzodisaster** [comment]
note: if I do the odd thing and append 'sess1' to the subject IDs in the inclusion list and pheno files, and try to run the group analysis CPAC errors out here:

> cpac_ga_model_generator.py", line 585, in prep_group_analysis_workflow
>     model_df["Filepath"][0]))[0])[0]).group(0))
> AttributeError: 'NoneType' object has no attribute 'group'
> found dead job  <Process(Process-7, stopped[1])>

**Comment by nimzodisaster** [comment]
I took some cases and I set up a new test site for individual analysis whose data was organized like ./<site>/<subject>/func/funcf.nii.gz
./<site>/<subject>/stuct/stuct.nii.gz

And the subject builde works just dandy now without putting it under a session directory.
OK, that is nice----but I've already preprocessed and QC my 300+ cases under ./<site>/<subject>/sess1/func/funcf.nii.gz organization as had been recommended here...

**Comment by nimzodisaster** [comment]
I took a subset of my processed subjects moved them to another directory. Using sed and rename, i remove all references to '_sess1' in directory and file names and within files. Then I retried my group analysis. The result is the same as my previous message with an error in cpac_ga_model_generator.py

The search string in the commented lines below e.g. sca_roi_(\d) failed to return anything. It appears that 1.0.0 that I had installed was naming with capital ROI, not roi. Like this

> /home/gose66/RS/out/pipeline_proc/111269-100_1_sess1/sca_roi_files_to_standard_fisher_zstd/_scan_func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score0/**sca_ROI_1_antswarp_fisher_zstd.nii.gz**

So I changed the regex

>         #out_dir = os.path.join(out_dir, \
>         #    re.search('sca_roi_(\d)+',os.path.splitext(\
>         #        os.path.splitext(os.path.basename(\
>         #            model_df["Filepath"][0]))[0])[0]).group(0))
>         #
>         # Change
>         out_dir = os.path.join(out_dir, \
>             re.search('sca_[Rr][Oo][Ii]_(\d)+',os.path.splitext(\
>                 os.path.splitext(os.path.basename(\
>                     model_df["Filepath"][0]))[0])[0]).group(0))

This change allowed the group analysis to continue, but then failed again...


> *+ WARNING: THD_open_tcat: should find wildcard or space in /home/gose66/RS/test/test2/out/pipeline_proc/111940-100/sca_roi_files_to_standard_fisher_zstd/_scan_func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2_antswarp_fisher_zstd.nii.gz
> *+ WARNING: THD_open_tcat: can't open dataset /home/gose66/RS/test/test2/out/pipeline_proc/111940-100/sca_roi_files_to_standard_fisher_zstd/_scan_func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2_antswarp_fisher_zstd.nii.gz[0] 
> *+ WARNING: THD_open_tcat: should find wildcard or space in /home/gose66/RS/test/test2/out/MODELS/group_analysis_results_proc/group_model_AmygdalaSCA0/sca_roi_files_to_standard_fisher_zstd/func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2/model_files/AmygdalaSCA0_sca_roi_files_to_standard_fisher_zstd_merged.nii.gz
> *+ WARNING: THD_open_tcat: can't open dataset /home/gose66/RS/test/test2/out/MODELS/group_analysis_results_proc/group_model_AmygdalaSCA0/sca_roi_files_to_standard_fisher_zstd/func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2/model_files/AmygdalaSCA0_sca_roi_files_to_standard_fisher_zstd_merged.nii.gz[0] 
> *** Failed to read catenation of /home/gose66/RS/test/test2/out/pipeline_proc/111940-100/sca_roi_files_to_standard_fisher_zstd/_scan_func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2_antswarp_fisher_zstd.nii.gz /home/gose66/RS/test/test2/out/MODELS/group_analysis_results_proc/group_model_AmygdalaSCA0/sca_roi_files_to_standard_fisher_zstd/func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2/model_files/AmygdalaSCA0_sca_roi_files_to_standard_fisher_zstd_merged.nii.gz[0] 
> Process Process-9:
> Traceback (most recent call last):
>   File "/home/gose66/miniconda/envs/cpac/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
>     self.run()
>   File "/home/gose66/miniconda/envs/cpac/lib/python2.7/multiprocessing/process.py", line 114, in run
>     self._target(*self._args, **self._kwargs)
>   File "/home/gose66/miniconda/envs/cpac/lib/python2.7/site-packages/CPAC/pipeline/cpac_ga_model_generator.py", line 878, in prep_group_analysis_workflow
>     check_merged_file(model_df["Filepath"], merge_file)
>   File "/home/gose66/miniconda/envs/cpac/lib/python2.7/site-packages/CPAC/pipeline/cpac_ga_model_generator.py", line 106, in check_merged_file
>     raise Exception(err)
> Exception: 
> 
> [!] Something went wrong while trying to run AFNI's 3ddot for the purpose of testing the merge file output.
> 
> Error details: Command '['3ddot', '-demean', '/home/gose66/RS/test/test2/out/pipeline_proc/111940-100/sca_roi_files_to_standard_fisher_zstd/_scan_func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2_antswarp_fisher_zstd.nii.gz', '/home/gose66/RS/test/test2/out/MODELS/group_analysis_results_proc/group_model_AmygdalaSCA0/sca_roi_files_to_standard_fisher_zstd/func/_threshold_0.25/_compcor_ncomponents_5_selector_pc10.linear0.wm1.global1.motion1.quadratic0.gm0.compcor0.csf1/_bandpass_freqs_0.008.0.08/_mask_LRamygdala_erode/_fisher_z_score1/sca_ROI_2/model_files/AmygdalaSCA0_sca_roi_files_to_standard_fisher_zstd_merged.nii.gz**[0]**']' returned non-zero exit status 1


Note the the [0] at the end of the file names. 

I am very frustrated.

**Comment by nimzodisaster** [comment]
> #test_string = ["3ddot", "-demean", output_file, merged_outfile + "[" + str(i) + "]"]
>         test_string = ["3ddot", "-demean", merged_outfile + "[" + str(i) + "]", output_file]

The commented line, taken from check_merged_file() in cpac_ga_model_generator.py was generating the error above. My version of 3ddot appears to require the merged volume to be listed as the first argument, not the second. I changed the code to the uncommented string above.
Finally, a group analysis ran...and completed.

Let me summarize the steps I had to make.

1. Because of a prior bug in the subject builder I had to place my subjec data under a session folder, despite having one session. This causes a mismatch in the Participant IDs held in the inclusion list and in new_df. To overcome this, I removes the session information from my directories and filenames, and within files using sed, grep, and rename utilities.
2. regex code in prep_group_analysis for sca_roi_ failed because my files were named  sca_ROI. I edited the regex to allow for both lower and upper case roi string.
3. input to 3ddot in check_merged_file() in cpac_ga_model_generator.py is entered in the incorrect order, at least for my version of 3ddot. I changed the order of inputs.

**Comment by florisvanvugt** [comment]
Thanks nimzodisaster, I had exactly the same issue and your fix solved it. I have incorporated it into my fork of the CPAC repo in case you want to pull from me. Let me know if you have further fixes.

**Comment by nimzodisaster** [comment]
You are most welcome, and I will definitely consider pulling from your repo as I am building another analysis workstation next week. Thanks!

**Comment by sgiavasis** [comment]
As mentioned on the [forum:](https://groups.google.com/forum/#!topic/cpax_forum/0UgBwyUPDmQ)

Okay, I managed to reproduce this error on an environment with the latest version of AFNI (17.2.02). I searched their release notes history to see if there's been some change, but I can't find anything. Although that code was written awhile ago, I can't really say for sure if AFNI's 3ddot was ever okay with having the sub-brick not being the first dataset input. Not sure how that didn't cause problems earlier, but I'd like to find out.

Going forward, the fix offered here will be included in the next point release. Thanks to everyone contributing.
